pkg/controllers/user/certsexpiration/certexpiration_test.go

```diff
@@ -0,0 +1,282 @@
+				"kube-etcd-172-17-0-4": v3.CertExpiration{},
+				"kube-etcd-172-17-0-5": v3.CertExpiration{},
+				"kube-node":            v3.CertExpiration{},
+				"kube-apiserver":       v3.CertExpiration{},
+				"kube-proxy":           v3.CertExpiration{},
+			},
+			cluster: &v3.Cluster{
+				Status: v3.ClusterStatus{
+					AppliedSpec: v3.ClusterSpec{
+						ClusterSpecBase: v3.ClusterSpecBase{
+							RancherKubernetesEngineConfig: &v3.RancherKubernetesEngineConfig{
+								Services: v3.RKEConfigServices{
+									Kubelet: v3.KubeletService{
+										GenerateServingCertificate: true,
+									},
+								},
+								Nodes: []v3.RKEConfigNode{
+									{
+										Address: "172.17.0.3",
+										Role: []string{
+											"etcd",
+										},
+									},
+									{
+										Address: "172.17.0.4",
+										Role: []string{
+											"etcd",
+										},
+									},
+									{
+										Address: "172.17.0.5",
+										Role: []string{
+											"etcd",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectNewCerts: map[string]v3.CertExpiration{
+				"kube-etcd-172-17-0-3": v3.CertExpiration{},
+				"kube-etcd-172-17-0-4": v3.CertExpiration{},
+				"kube-etcd-172-17-0-5": v3.CertExpiration{},
+				"kube-node":            v3.CertExpiration{},
+				"kube-apiserver":       v3.CertExpiration{},
+				"kube-proxy":           v3.CertExpiration{},
+			},
+		},
+		{
+			name: "Keep valid kubelet certs",
+			certs: map[string]v3.CertExpiration{
+				"kube-node":               v3.CertExpiration{},
+				"kube-kubelet-172-17-0-4": v3.CertExpiration{},
+				"kube-kubelet-172-17-0-3": v3.CertExpiration{},
+				"kube-kubelet-172-17-0-5": v3.CertExpiration{},
+				"kube-etcd-172-17-0-5":    v3.CertExpiration{},
+				"kube-apiserver":          v3.CertExpiration{},
+				"kube-proxy":              v3.CertExpiration{},
+			},
+			cluster: &v3.Cluster{
+				Status: v3.ClusterStatus{
+					AppliedSpec: v3.ClusterSpec{
+						ClusterSpecBase: v3.ClusterSpecBase{
+							RancherKubernetesEngineConfig: &v3.RancherKubernetesEngineConfig{
+								Services: v3.RKEConfigServices{
+									Kubelet: v3.KubeletService{
+										GenerateServingCertificate: true,
+									},
+								},
+								Nodes: []v3.RKEConfigNode{
+									{
+										Address: "172.17.0.3",
+										Role: []string{
+											"worker",
+										},
+									},
+									{
+										Address: "172.17.0.4",
+										Role: []string{
+											"worker",
+										},
+									},
+									{
+										Address: "172.17.0.5",
+										Role: []string{
+											"etcd",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectNewCerts: map[string]v3.CertExpiration{
+				"kube-node":               v3.CertExpiration{},
+				"kube-kubelet-172-17-0-4": v3.CertExpiration{},
+				"kube-kubelet-172-17-0-3": v3.CertExpiration{},
+				"kube-kubelet-172-17-0-5": v3.CertExpiration{},
```

Comment:
Why is this kubelet cert present when `172.17.0.5` is an etcd only node?

Reply:
When  generating kubelet certs,all hosts be used:
https://github.com/rancher/rke/blob/18548be050e7badc15dd31f9332aea80cc4db3ab/pki/services.go#L495